### ISSUE TITLE
roachperf: regression around 2022-08-10

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#86541]
[`kv95/enc=false/nodes=3/cpu=32`](https://roachperf.crdb.dev/?filter=&view=kv95%2Fenc%3Dfalse%2Fnodes%3D3%2Fcpu%3D32&tab=gce) on GCE shows a regression around 2022-08-10:

<img width="780" alt="Screenshot 2022-08-22 at 10 23 27" src="https://user-images.githubusercontent.com/644420/185874776-20e24dfd-5fc9-4553-9ff4-7bcc0e02ae9d.png">

Jira issue: CRDB-18803

**Comment by erikgrinaker** [comment]
@nvanbenschoten Bisected this down to c7c154af8c (#85764). Raw notes

```
./workload run kv --init --splits 1000 --concurrency 192 --read-percent 95 --ramp 120s --duration 180s

Random samples:
dc32c9248e 2022-08-08: vendor: bump Pebble to 1afadc20bb70 [Jackson Owens] 117608.7
d210d76a55 2022-08-09: vendor: bump Pebble to cb25d247e7c2 [Jackson Owens] 116327.9
9fb8d361ec 2022-08-11: Merge #84107 #85580 #85778 #85943 #85951 [craig[bot]] 111666.1 116822.9
fb87e5f54c 2022-08-11: vendor: bump Pebble to c06e0793dbf5; mark pre-Pebblev1 tables [Nick Travers] 114533.6 117155.0
f1c20476ed 2022-08-12: Merge #85095 #85777 [craig[bot]] 113543.3
d94f32c4d9 2022-08-12: vendor: bump Pebble to 84fa6a019ed4 [Jackson Owens] 113493.8 113414.6
e6a7dc2f8e 2022-08-21: Merge #86174 [craig[bot]] (HEAD -> master) 112350.3

Bisecting fb87e5f54c (good) to f1c20476ed (bad)
f4072978c8 2022-08-12: Merge #85826 #85998 [craig[bot]] 112278.1
468ac0188d 2022-08-12: Merge #82477 [craig[bot]] 117451.2
1ac67a9d10 2022-08-12: Merge #85764 [craig[bot]] 114363.7 113268.3
c4bf42a770 2022-08-08: sql: make mutations visible to volatile UDFs [Marcus Gartner] 116228.7
e17eb3649f 2022-08-12: Merge #85722 #85786 [craig[bot]] 116137.5
c7c154af8c 2022-08-08: kv: pass explicit Now timestamp on BatchRequest, remove timestamp downcasting [Nathan VanBenschoten] 110963.6

c7c154af8cf34c69aff26ce2b8e985fa8fd2e891 is the first bad commit
commit c7c154af8cf34c69aff26ce2b8e985fa8fd2e891
Author: Nathan VanBenschoten <nvanbenschoten@gmail.com>
Date:   Mon Aug 8 14:47:55 2022 -0400

    kv: pass explicit Now timestamp on BatchRequest, remove timestamp downcasting
```

**Comment by nvanbenschoten** [comment]
Thanks for flagging this @erikgrinaker.

Since we're seeing this on 32 vCPU machines (2 numa nodes), I suspect that this is an artifact of the additional exclusive mutex acquisition in the HLC when grabbing a timestamp.

One thing that we can explore is probabilistically removing this HLC access. In fact, there are a few places where we can now do something like this, now that these clock accesses are strictly for HLC stabilization and not necessary for transaction consistency:
https://github.com/cockroachdb/cockroach/blob/9c7db33c1ee21145fe372d61ad26b45a9f151ee3/pkg/kv/kvclient/kvcoord/dist_sender.go#L641-L643

**Comment by erikgrinaker** [comment]
Definitely seeing it on 8-core boxes too:

<img width="799" alt="Screenshot 2022-08-22 at 18 50 11" src="https://user-images.githubusercontent.com/644420/185975798-d5ee0279-f81e-4d65-bb30-ba418a45e297.png">

**Comment by srosenberg** [comment]
> Thanks for flagging this @erikgrinaker.
> 
> Since we're seeing this on 32 vCPU machines (2 numa nodes), I suspect that this is an artifact of the additional exclusive mutex acquisition in the HLC when grabbing a timestamp.
> 
> One thing that we can explore is probabilistically removing this HLC access. In fact, there are a few places where we can now do something like this, now that these clock accesses are strictly for HLC stabilization and not necessary for transaction consistency:
> 
> https://github.com/cockroachdb/cockroach/blob/9c7db33c1ee21145fe372d61ad26b45a9f151ee3/pkg/kv/kvclient/kvcoord/dist_sender.go#L641-L643

I wonder if the body could be rewritten without the mutex while preserving the invariant: `(WallTime == physicalClock && Logical == 0) || (WallTime >= physicalClock && next(Logical) >= Logical + 1)`? The postcondition, `enforceWallTimeWithinBoundLocked`, doesn't require a mutex either.

```
next := atomic.AddInt32(&c.mu.timestamp.Logical, 1)

for ;c.mu.timestamp.WallTime < physicalClock; {
    atomic.StoreInt64(&c.mu.timestamp.WallTime, physicalClock)
    atomic.CompareAndSwapInt32(&c.mu.timestamp.Logical, next+int32(c.mu.timestamp.WallTime-physicalClock), 0)
}
```

**Comment by nvanbenschoten** [comment]
I ran a series of experiments to determine where the regression was coming from in this commit. If it had been the extra HLC access, I think a change in the direction of what @srosenberg was suggesting makes sense.

However, it doesn't seem to be due to the added HLC access. I bisected the new code in the commit to the proto change. If I add the new field to the BatchRequest header, even without touching it, I get the regression. Furthermore, even if I make the field nullable so that it has zero cost over the wire, I still get the regression.

I just checked the `sizeof(BatchRequest)` and see that it is `264` bytes with this new field. That's just above 256 `bytes`. I wonder if that means we now pull heap allocated BatchRequests from the 512 byte [small object size class](https://github.com/golang/go/blob/ba5deb408f50f31a872f114c5ef7b52090e4b753/src/runtime/malloc.go#L10).
